document_stores/elasticsearch/src/elasticsearch_haystack/document_store.py

```diff
@@ -0,0 +1,305 @@
+             - fail: an error is raised
+        :raises DuplicateDocumentError: Exception trigger on duplicate document if `policy=DuplicatePolicy.FAIL`
+        :return: None
+        """
+        if len(documents) > 0:
+            if not isinstance(documents[0], Document):
+                msg = "param 'documents' must contain a list of objects of type Document"
+                raise ValueError(msg)
+
+        action = "index" if policy == DuplicatePolicy.OVERWRITE else "create"
+        _, errors = helpers.bulk(
+            client=self._client,
+            actions=(
+                {
+                    "_op_type": action,
+                    "_id": doc.id,
+                    "_source": doc.to_dict(),
+                }
+                for doc in documents
+            ),
+            refresh="wait_for",
+            index=self._index,
+            raise_on_error=False,
+        )
+        if errors and policy == DuplicatePolicy.FAIL:
+            # TODO: Handle errors in a better way, we're assuming that all errors
+            # are related to duplicate documents but that could be very well be wrong.
+
+            # mypy complains that `errors`` could be either `int` or a `list` of `dict`s.
+            # Since the type depends on the parameters passed to `helpers.bulk()`` we know
+            # for sure that it will be a `list`.
+            ids = ", ".join(e["create"]["_id"] for e in errors)  # type: ignore[union-attr]
+            msg = f"IDs '{ids}' already exist in the document store."
+            raise DuplicateDocumentError(msg)
+
+    def _deserialize_document(self, hit: Dict[str, Any]) -> Document:
+        """
+        Creates a Document from the search hit provided.
+        This is mostly useful in self.filter_documents().
+        """
+        data = hit["_source"]
+
+        if "highlight" in hit:
+            data["metadata"]["highlighted"] = hit["highlight"]
+        data["score"] = hit["_score"]
+
+        return Document.from_dict(data)
+
+    def delete_documents(self, document_ids: List[str]) -> None:
+        """
+        Deletes all documents with a matching document_ids from the document store.
+
+        :param object_ids: the object_ids to delete
+        """
+
+        #
+        helpers.bulk(
+            client=self._client,
+            actions=({"_op_type": "delete", "_id": id_} for id_ in document_ids),
+            refresh="wait_for",
+            index=self._index,
+            raise_on_error=False,
+        )
+
+    def _bm25_retrieval(
+        self,
+        query: str,
+        *,
+        filters: Optional[Dict[str, Any]] = None,
+        top_k: int = 10,
+        scale_score: bool = True,
+    ) -> List[Document]:
+        """
+        Elasticsearch by defaults uses BM25 search algorithm.
+        Even though this method is called `bm25_retrieval` it searches for `query`
+        using the search algorithm `_client` was configured with.
```

Comment:
I struggle to understand this part. How can I configure a search algorithm for the `_client`?

Reply:
It depends on the similary settings of the index used. 
If the user creates the Document Store using an existing index that has been configured with different settings it would use those. 
If the index doesn't exist and is created when the Document Store is instantiated the default will be BM25 so it's not an issue. There is no way as of now to create the index with different settings when creating the Document Store.

There's some more info in the [official docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-similarity.html) about the different similarities supported.